Objects/unicodeobject.c

```diff
@@ -2217,6 +2226,58 @@ PyUnicode_FromWideChar(const wchar_t *u, Py_ssize_t size)
     if (size == 0)
         _Py_RETURN_UNICODE_EMPTY();
 
+#if defined(__sun) && defined(__SVR4)
+    /* Check whether current locale uses UTF to encode symbols */
+    if (!codeset_is_utf8_compatible()) {
+
+        /* Given 'u' might not be NULL terminated (size smaller than its
+           length); copy and terminate part we are interested in. */
+        wchar_t* substr = PyMem_RawMalloc((size + 1) * sizeof(wchar_t));
```

Comment:
Please check for integer overflow. Example:

```
    if (length > ((PY_SSIZE_T_MAX - struct_size) / char_size - 1)) {
        PyErr_NoMemory();
        return NULL;
    }
    new_size = (struct_size + (length + 1) * char_size);
```

Same on the second PyMem_RawMalloc() call below.

Since we own the GIL, I would prefer to use PyMem_Malloc() which can be faster (use pymalloc): https://docs.python.org/dev/c-api/memory.html

Reply:
I added checks for both `size` overflow and successful allocation, and changed all allocations to `PyMem_Malloc()`.